docs(sdk): Update README.md

ngrayluna opened the pull request:
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
First draft of new README.

Questions:
-----------
1. What picture to use below the W&B logo? Do we have a nice-looking, informative image we like?
2. Whichever image we use, it would be could to make the images click-able; when selected they redirect you to said page info. e.g. See what Yolov5 did in their environments section: https://github.com/ultralytics/yolov5#environments
3. Where do we store new images? In [./.github](https://github.com/wandb/wandb/tree/main/.github) or maybe [wandb/assets](https://github.com/wandb/assets)?

Testing
-------
See my forked repo's update_README for preview of the new README draft: https://github.com/ngrayluna/wandb/tree/update_README

Checklist
-------
- [ ] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/master/CONTRIBUTING.md#conventional-commits)

sephmard commented on README.md:
```diff
@@ -5,37 +5,42 @@
+</p>
 
-## Features
+[It would be cool to have make the above icons click-able. Thoughts?]
```
Yes! I've seen this pattern in other github repos and think its a polished feel

scottire commented on README.md:
```diff
@@ -5,37 +5,42 @@
-- Focus on the interesting ML. Spend less time manually tracking results in spreadsheets and text files.
-- Capture dataset versions with W&B Artifacts to identify how changing data affects your resulting models.
-- Reproduce any model, with saved code, hyperparameters, launch commands, input data, and resulting model weights.
+[Add more info, less markety sentence or two here?]
```
Giving people a bunch of links to our best examples/docs in a table would be useful. Maybe embedding our product overview video would be useful.

Example useful links:
Track your models metrics and configs with live dashboards (link getting started)
Version datasets and models with Artifacts and Model Registry (links)
Tune hyperparameters with Sweeps (links)
Understand model predictions with interactive Tables

morganmcg1 replied:
I like Scott's suggestions

> Version datasets and models with Artifacts and Model Registry
I think  we need to drop talk of Artifacts and data versioning and focus on Model Checkpointing and Model Registry

morganmcg1 replied:
> Understand model predictions with interactive Tables

Maybe drop Tables also, 3 is a nice number of bullet points :D

morganmcg1 reviewed (changes requested):
Can we make sure all links have utms please

morganmcg1 commented:
I think a mention & link to W&B Server is important to highlight, with a couple of quick value props (local, private etc) https://docs.wandb.ai/guides/self-hosted - @sephmard can message here

Possibly mentioning that we work in a distributed setting too as another value prop. Its a question that comes up fairly often, https://docs.wandb.ai/guides/track/advanced/distributed-training - this one I'm less sure about

dmitryduev commented on README.md:
```diff
@@ -192,126 +174,22 @@ trainer = Trainer(logger=wandb_logger)
-Just run a script using HuggingFace's Trainer passing `--report_to wandb` to it
-in an environment where `wandb` is installed, and we'll automatically log losses,
-evaluation metrics, model topology, and gradients:
+&nbsp;
```
Mention + add a link to a doc where all our integrations are mentioned? We have a bunch...

dmitryduev replied:
Keep HuggingFace?

dmitryduev commented on README.md:
````diff
@@ -192,126 +174,22 @@ trainer = Trainer(logger=wandb_logger)
-docker run --rm -d -v wandb:/vol -p 8080:8080 --name wandb-local wandb/local
-```
+# Contact
+If you have any questions, please don't hesitate to ask in our [user forum](http://wandb.me/forum). For wandb bugs and feature requests please visit [GitHub Issues](https://github.com/wandb/wandb/issues). For professional support please Contact Us. Be a part of the W&B Community, post your qeustions at [W&B Community](https://community.wandb.ai/). Stay connected with the latest updates with [W&B  ...
````
```suggestion
If you have any questions, please don't hesitate to ask in our [user forum](http://wandb.me/forum). For wandb bugs and feature requests please visit [GitHub Issues](https://github.com/wandb/wandb/issues). For professional support please Contact Us. Be a part of the W&B Community, post your questions at [W&B Community](https://community.wandb.ai/). Stay connected with the latest updates with [W&B Fully Connected](https://wandb.ai/fully-connected).
```

dmitryduev replied:
Contact Us goes to https://wandb.ai/site/contact?

ngrayluna replied:
Oops. I'll add this in the next PR.

dmitryduev reviewed (approved):
Love it, great work @ngrayluna, thanks a ton! I second essentially everything that people have said so far + added a few nits.

ngrayluna commented on README.md:
```diff
@@ -161,157 +215,60 @@ with tf.Session() as sess:
+# Try out our MLOps Course
 
-<img src="https://i.imgur.com/THYXBN0.png" width="720" alt="Weights & Biases" />
+Get MLOps Certified With The Course From Weights & Biases.
```
Do we want to link to the MLOps course?

dmitryduev replied:
@morganmcg1? I don't mind, it's an awesome course!

ngrayluna commented on README.md:
````diff
@@ -161,157 +215,60 @@ with tf.Session() as sess:
-```shell
-# 1. Install the wandb library
-pip install wandb
+# Tutorials
````
Which Tutorials do we want to highlight here considering our target audience?

dmitryduev replied:
@morganmcg1 could you provide a list please?

ngrayluna replied:
@morganmcg1 Do you have a list? If not, we can omit this part for now and add it later.

ngrayluna commented on README.md:
```diff
@@ -3,90 +3,107 @@
-- **[See how to set configs in a colab →](http://wandb.me/config-colab)**
-- [Docs](https://docs.wandb.com/library/config)
+<p align='center'>
+<img src="./README_images/integrations.png" width="100%" />
```
Not a huge fan of this images...wondering if we have a more stylish looking images of our integrations?

dmitryduev replied:
+1!

dmitryduev replied:
i'd also use a version of this (or other) png with a transparent bg so that in dark mode it looks as nice as on our marketing site: 
<img width="1248" alt="image" src="https://user-images.githubusercontent.com/7557205/210480514-95da16cf-45dd-458a-9671-1f3b07825d5a.png">
(https://assets.website-files.com/5ac6b7f2924c656f2b13a88c/637d1468166be22fd83e4155_integrations-transparent-wanbd.png)

ngrayluna replied:
Waiting for design to reply....

Might vote to keep image as is for now, update when we have a new image to use.

ngrayluna replied:
Design team replied, their input is:
1. We shouldn't have frameworks displayed as large as they are now.
2. We should match the frameworks we show with the examples we have on the README.

So as to not be blocked, I have removed the image with all the frameworks.

ngrayluna commented on README.md:
```diff
@@ -3,90 +3,107 @@
-- Focus on the interesting ML. Spend less time manually tracking results in spreadsheets and text files.
-- Capture dataset versions with W&B Artifacts to identify how changing data affects your resulting models.
-- Reproduce any model, with saved code, hyperparameters, launch commands, input data, and resulting model weights.
+<p align='center'>
```
There are placeholder images, waiting to hear from design if they have higher quality assets and assets w/ clear backgrounds.

dmitryduev replied:
still, love how these look! in case things stay as is, here's a tiny size mismatch: 
<img width="448" alt="image" src="https://user-images.githubusercontent.com/7557205/210479501-6bad8ab0-1f70-441a-9bcd-dd861c214bc9.png">

ngrayluna commented on README.md:
```diff
@@ -3,90 +3,107 @@
+<a href="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/intro/Intro_to_Weights_%26_Biases.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" /></a>
+</p>
 
 Use W&B to build better models faster. Track and visualize all the pieces of your machine learning pipeline, from datasets to production models.
```
With the images pointing to the docs, do we want to still highlight/callout three product areas (eg Experiments)? 

If nay, any thoughts on what else we should add on the opening paragraph?

dmitryduev replied:
I vote nay to calling those out -- all the product areas are now accessible via the SDK (including the Reports, for example). Love the opening paragraph. Nit: it's obvious, but I'd say "machine learning models" instead of just "models".

ngrayluna commented on README.md:
````diff
@@ -3,90 +3,107 @@
 ```
 
-### [Try in a colab →](http://wandb.me/intro)
+3.5. Use the example code snippet below as a template to integrate W&B to your Python script:
````
Typo: Change 3.5. to 3

dmitryduev replied:
nvm - I pushed fixes to all replacement typos I could find

dmitryduev commented on README.md:
```diff
@@ -3,90 +3,107 @@
+<img src="./README_images/wandb_demo_experiments.gif" width="100%">
+</p>
+<p align = "center">
+Example W&B Dashboard that shows Wuns from an Experiment.
```
typo (`Wuns`)

dmitryduev commented on README.md:
```diff
@@ -101,7 +118,7 @@ config.learning_rate = 0.01
 # Model training code here ...
 
-# Step 3: Add WandbCallback
+# Step 3.5: Add WandbCallback
```
```suggestion
# Step 3: Add WandbCallback
```

dmitryduev commented on README.md:
```diff
@@ -137,22 +187,26 @@ for batch_idx, (data, target) in enumerate(train_loader):
 
 # 1. Start a W&B run
-wandb.init(project="gpt3")
+wandb.init(project="gpt3.5")
```
```suggestion
wandb.init(project="gpt4")
```

dmitryduev commented on README.md:
```diff
@@ -137,22 +187,26 @@ for batch_idx, (data, target) in enumerate(train_loader):
 # Model training here
 
-# 3. Log metrics over time to visualize performance
+# 3.5. Log metrics over time to visualize performance
```
```suggestion
# 3. Log metrics over time to visualize performance
```

dmitryduev commented on README.md:
```diff
@@ -161,157 +215,60 @@ with tf.Session() as sess:
 from pytorch_lightning import Trainer
 
-wandb_logger = WandbLogger(project="gpt-3")
+wandb_logger = WandbLogger(project="gpt-3.5")
```
```suggestion
wandb_logger = WandbLogger(project="gpt4")
```

dmitryduev replied:
let's be consistently outlooking :)

dmitryduev commented on README.md:
```diff
@@ -3,90 +3,107 @@
 
+# Model training code here ...
 
-# Academic Researchers
```
I think we should keep this section (albeit with modifications -- mirroring and compressing https://wandb.ai/site/research) or at least highlighted that W&B is "Free for academic and open source projects."

dmitryduev commented on README.md:
```diff
@@ -111,20 +128,53 @@ model.fit(
 
 # 1. Start a new run
-wandb.init(project="gpt-3")
+wandb.init(project="gpt-3.5")
```
```suggestion
wandb.init(project="gpt4")
```

dmitryduev commented on README.md:
```diff
@@ -111,20 +128,53 @@ model.fit(
 config.dropout = 0.01
 
-# 3. Log gradients and model parameters
+# 3.5. Log gradients and model parameters
```
```suggestion
# 3. Log gradients and model parameters
```

dmitryduev commented on README.md:
```diff
@@ -3,90 +3,107 @@
-- Capture dataset versions with W&B Artifacts to identify how changing data affects your resulting models.
-- Reproduce any model, with saved code, hyperparameters, launch commands, input data, and resulting model weights.
+<p align='center'>
+<a href="https://docs.wandb.ai/guides/track"><img src="./README_images/experiments.png" width="13.5%" /></a>
```
Please move the image assets to ./doc/README_images (we're trying to keep the repo root level as clean as possible)

dmitryduev reviewed (changes requested):
❤️

morganmcg1 commented:
@ngrayluna @dmitryduev we've just updated the Quickstart in the app that users land on. It includes end to end runnable examples for 6 libraries

The list of libraries and runnable code for each [are all here](https://www.notion.so/wandbai/Quickstart-Integrations-Code-fa8a2e48bdf045c0b56fa53081f89d13), it'd be great to replace the code in the current Integrations sections with these

dmitryduev commented on README.md:
```diff
@@ -161,157 +215,60 @@ with tf.Session() as sess:
-**Learn about Artifacts [here →](https://www.wandb.com/articles/announcing-artifacts) | Read the [Docs](https://docs.wandb.com/artifacts)**
+Bringing machine learning models to production is challenging, with a continuous iterative lifecycle that consists of many complex components. Having a disciplined, flexible and collaborative process - an effective MLOps system - is crucial to enabling velocity and rigor, and building an end-to-end machine learning pipeline that continually delivers production-ready ML models and services.
 
-# 💻  Run W&B Server Locally
```
I would also keep this section (maybe in a shortened form) -- many people seems not to know that this it is possible to run server locally.

ngrayluna replied:
Oops. Missed this. I'll add it in the next PR.

ngrayluna commented:
> @ngrayluna @dmitryduev we've just updated the Quickstart in the app that users land on. It includes end to end runnable examples for 6 libraries
> 
> The list of libraries and runnable code for each [are all here](https://www.notion.so/wandbai/Quickstart-Integrations-Code-fa8a2e48bdf045c0b56fa53081f89d13), it'd be great to replace the code in the current Integrations sections with these

Updated the Keras, Lightning, and HG examples. Didn't update Torch example since it doesn't use wandb.log .

ngrayluna commented:
Added a Server section. It's basic at the moment. We are working to revamp those docs so I don't want to add conflicting info.

dmitryduev reviewed (approved):
Absolutely love it, awesome work @ngrayluna!

One last thing and should be ready to be merged, as far as I'm concerned:
<img width="776" alt="image" src="https://user-images.githubusercontent.com/7557205/211125769-1501fc91-813f-4fa4-b7b8-a3dc0c09d812.png">
- Launch and Models icons have a non-transparent bg
- The fonts are hard to read on GH when using a dark theme, I wonder if there's a font color that works both for light and dark themes?

morganmcg1 commented on README.md:
````diff
@@ -52,40 +57,24 @@ for i in range(10):
-wandb.config.architecture = "resnet"
-```
+# Integrations
+Use your favorite framework with W&B. W&B integrations make it fast and easy to set up experiment tracking and data versioning inside existing projects.
````
@ngrayluna can we add this docs link please?

morganmcg1 replied:
Also few more changes to the Integrations list

Can you change the list to:

🔥 PyTorch
🌊 TensorFlow / Keras
🤗 Hugging Face Transformers
⚡️ PyTorch Lightning
💨 XGBoost
🧮 Sci-Kit Learn

And add code for each of them from [this notion doc](https://www.notion.so/wandbai/Quickstart-Integrations-Code-fa8a2e48bdf045c0b56fa53081f89d13) please?

Changes:
- Tensorflow and Keras can be merged as modern TF mostly relies on Keras now
- Huggingface renamed to Huggingface Transformers as HF now have many many libraries
- Added XGBoost and Sci-Kit Learn as its important to show wand also works for non deeplearning libraries too

ngrayluna replied:
Fixed.
